src/v/storage/disk_log_impl.cc

```diff
@@ -2516,8 +2522,17 @@ disk_log_impl::remove_prefix_full_segments(truncate_prefix_config cfg) {
       },
       [this] {
           auto ptr = _segs.front();
+          /**
+           * If segment has outstanding locks wait for it to be unlocked before
+           * prefixing truncating it, this way the prefix truncation will
+           * naturally yield to the appender and it will be retried
+           */
+          if (ptr->has_outstanding_locks()) {
```

Comment:
remove_segment_permanently() already does this with an info log, perhaps move this sleep there?

a question for my education, I see has_outstanding_locks() doesn't check for `_segments_rolling_lock` and `_segment_rewrite_lock`, is it safe to remove segments while while rolling in progress? It probably is, I didn't fully wrap my head around the locking protocol here.

Reply:
Staring more at the code, IMO removing the segment from the _segs map before doing anything seems a bit sketch and confusing to me, I was wondering if that should be done inside remove_segment_permanently() after all the initial checks pass?

Reply:
> removing the segment from the _segs map before doing anything seems a bit sketch and confusing to me

+1 to that, if we are adding `is_tombstone()` checks in all relevant places, maybe we don't need to remove right away